src/Interval.h

```diff
@@ -0,0 +1,301 @@
+
+#include <algorithm>
+#include <ostream>
+
+#include "SpatialRelation.h"
+
+
+namespace lsst {
+namespace sg {
+
+/// `Interval` represents a closed interval of the real numbers
+/// by its upper and lower bounds. It is parameterized by the Scalar
+/// type, which must be constructible from a double.
+///
+/// An interval with identical upper and lower bounds contains
+/// a single point, and is equal to that scalar bound.
+///
+/// An interval with an upper bound less than its lower bound is
+/// empty (contains no points), as are intervals with NaN as
+/// either bound.
+template <typename Derived, typename Scalar>
+class Interval {
+public:
+    /// This constructor creates an empty interval.
+    Interval() : _a(1.0), _b(0.0) {}
+    /// This constructor creates a closed interval containing only x.
+    explicit Interval(Scalar x) : _a(x), _b(x) {}
+    /// This constructor creates an interval from the given endpoints.
+    Interval(Scalar x, Scalar y) : _a(x), _b(y) {}
+
+    /// Two closed intervals are equal if their endpoints are the same, or
+    /// both are empty.
+    bool operator==(Interval const & i) const {
+        return (_a == i._a && _b == i._b) || (i.isEmpty() && isEmpty());
+    }
+    bool operator!=(Interval const & i) const { return !(*this == i); }
+
+    /// A closed interval is equal to a point x if both endpoints equal x.
+    bool operator==(Scalar x) const {
+        return (_a == x && _b == x) || (x != x && isEmpty());
+    }
+    bool operator!=(Scalar x) const { return !(*this == x); }
+
+    ///@{
+    /// `a` returns the lower endpoint of this interval. The return value
+    /// for empty intervals is arbitrary.
+    Scalar a() const { return _a; }
+    Scalar & a() { return _a; }
+    ///@}
+
+    ///@{
+    /// `b` returns the upper endpoint of this interval. The return value
+    /// for empty intervals is arbitrary.
+    Scalar b() const { return _b; }
+    Scalar & b() { return _b; }
+    ///@}
+
+    /// `isEmpty` returns true if this interval does not contain any points.
+    bool isEmpty() const {
+        return !(_a <= _b); // returns true when _a and/or _b is NaN
+    }
+
+    /// `center` returns the center of this interval. It is arbitrary
+    /// for empty intervals.
+    Scalar center() const { return 0.5 * (_a + _b); }
+    /// `size` returns the size (length, width) of this interval. It is zero
+    /// for single-point intervals, and NaN or negative for empty intervals.
+    Scalar size() const { return _b - _a; }
+
+    ///@{
+    /// `contains` returns true if the intersection of this interval and x
+    /// is equal to x.
+    bool contains(Scalar x) const {
+        return (_a <= x && x <= _b) || x != x;
+    }
+    bool contains(Interval const & x) const {
+        if (x.isEmpty()) {
+            return true;
+        } else if (isEmpty()) {
+            return false;
+        }
+        return _a <= x._a && _b >= x._b;
+    }
+    ///@}
```

Comment:
I'm honestly not sure what's more useful in these edge cases - I went with the simple definition because it made it easier for me to figure out how the code should behave.